File: sem3/PP-scala/laby3/src/main/scala/zad4.scala
object zad4 {
  def reverse_list[T](list: List[T]): List[T] = {
    list match {
      case Nil => Nil
      case x :: rest => reverse_list(rest) ::: List(x)
    }
  }

  def main(args: Array[String]): Unit = {
    val list1 = List(1, 2, 3, 4, 5)
    val list2 = List("a", "b", "c", "d")

    println(reverse_list(list1))
    println(reverse_list(list2))
  }
}

File: sem3/PP-ocaml/lab3/zad4/zad4.ml
let rec reverse_list xs =
  let rec reverse_pom ys output =
    match ys with
    | [] -> output
    | x :: rest -> reverse_pom rest (x :: output)
  in reverse_pom xs [];;

let rec print_list lista =
  match lista with
  | [] -> ()
  | x :: reszta ->
    print_int x;
    print_string " ";
    print_list reszta;;

let moja_lista = [1; 2; 3; 4; 5];;
let lista_odwrotna = reverse_list moja_lista;;

print_list moja_lista;;
print_newline ();;
print_list lista_odwrotna;;
print_newline ();;